table/column.go

```diff
@@ -191,6 +191,34 @@ func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo, r
 		truncateTrailingSpaces(&casted)
 	}
 
+	if (sc.InInsertStmt || sc.InUpdateStmt) && !casted.IsNull() && (col.Tp == mysql.TypeDate || col.Tp == mysql.TypeDatetime) {
+		tm := casted.GetMysqlTime()
+		mode := ctx.GetSessionVars().SQLMode
+		ignoreErr := sc.DupKeyAsWarning
+
+		if (mode.HasNoZeroDateMode() && tm.IsZero()) ||
+			(mode.HasNoZeroInDateMode() && tm.Year() > 0 && (tm.Month() == 0 || tm.Day() == 0)) {
+			if mode.HasStrictMode() && !ignoreErr {
+				return casted, types.ErrWrongValue.GenWithStackByArgs(types.DateTimeStr, tm.String())
+			}
+			var (
+				zeroV interface{}
```

Comment:
Could you please use `types.Time` instead of `interface{}`?

Reply:
+1